### PR TITLE
Fix `#[extendr]` conversions into `Dataframe<T>`

### PR DESCRIPTION
**Pull request by CGMossa**
Addressing #715

### CODE REVIEW

**Comment by CGMossa** [comment]
This issue has two parts. It is not possible to use it like so:

```rs
#[derive(IntoDataFrameRow)]
struct Row {
    name: u32,
}

#[extendr]
fn dataframe_conversion(data_frame: Dataframe<Row>) -> Robj {
    vec![Row { name: 42 }].into_dataframe().unwrap().into_robj()
}
```

File: extendr-api/src/wrapper/dataframe.rs [hunk 1]
```diff
@@ -7,7 +7,25 @@ pub trait IntoDataFrameRow<T> {
...
+    _marker: std::marker::PhantomData<T>,
+}
+
+impl<T> From<Dataframe<T>> for Robj {
+    fn from(value: Dataframe<T>) -> Self {
+        value.robj
+    }
+}
+
+impl<T> FromRobj<'_> for Dataframe<T> {
```
Comment (Ilia-Kosenkov) [inline extendr-api/src/wrapper/dataframe.rs:19]: Can you delegate this impl to `TryFrom`?

File: extendr-api/tests/dataframe_tests.rs [hunk 2]
```diff
@@ -51,3 +51,39 @@ fn test_into_robj_dataframe() {
...
+    impl Row {}
+
+    #[extendr]
+    fn use_dataframe_exptr(data_frame: Dataframe<Row>) -> Dataframe<Row> {
+        data_frame
+    }
+}
+
+#[test]
+fn test_dataframe_row_and_externalptr() {
```
Comment (Ilia-Kosenkov) [inline extendr-api/tests/dataframe_tests.rs:71]: These tests do not really test anything other than code generation. I'd make them 'integration' tests.

File: extendr-api/src/wrapper/dataframe.rs (hunk 1 above)
Comment (CGMossa) [reply extendr-api/src/wrapper/dataframe.rs:19]: It is already available for TryFrom. I'm expanding the compatibility, until we drop the big Only-TryFrom PR.

File: extendr-api/src/wrapper/dataframe.rs [hunk 3]
```diff
@@ -63,8 +81,33 @@ where
...
+    use super::*;
+    use crate as extendr_api;
+
+    #[derive(IntoDataFrameRow)]
+    struct Row {
+        name: u32,
+    }
+
+    #[extendr]
+    fn dataframe_conversion(_data_frame: Dataframe<Row>) -> Robj {
```
[the remark this answers is not in the document]
Comment (CGMossa) [reply extendr-api/src/wrapper/dataframe.rs:95]: These didn't compile before.

File: extendr-api/tests/dataframe_tests.rs (hunk 2 above)
Comment (CGMossa) [reply extendr-api/tests/dataframe_tests.rs:71]: They are already in integration tests, as they are stored in `extendr-api/tests` folder.

**Comment by CGMossa** [comment]
Please take a look again, as this PR fixes things that didn't work before. See the linked issue.

**Comment by CGMossa** [comment]
@Ilia-Kosenkov

File: extendr-api/src/wrapper/dataframe.rs (hunk 1 above)
Comment (Ilia-Kosenkov) [reply extendr-api/src/wrapper/dataframe.rs:19]: I meant you duplicate coed here and in TryFrom

File: extendr-api/tests/dataframe_tests.rs (hunk 2 above)
Comment (Ilia-Kosenkov) [reply extendr-api/tests/dataframe_tests.rs:71]: I meant on R side. Here these functions just 'compile', there are no real 'tests' her IMO.

**Comment by Ilia-Kosenkov** [comment]
Here, I found how you can test that these functions actually work as intended, including conversions
https://github.com/extendr/extendr/blob/e61fc0a6c9e963e08df0f39a3bddb9a48bdacfe4/extendr-api/src/lib.rs#L815-L864

File: extendr-api/src/wrapper/dataframe.rs (hunk 1 above)
Comment (CGMossa) [reply extendr-api/src/wrapper/dataframe.rs:19]: Good hint. Hold on.

File: extendr-api/tests/dataframe_tests.rs (hunk 2 above)
Comment (CGMossa) [reply extendr-api/tests/dataframe_tests.rs:71]: They didn't used to compile. They compile now. I'm not going to add more code here.

**Comment by CGMossa** [comment]
> Here, I found how you can test that these functions actually work as intended, including conversions
> 
> https://github.com/extendr/extendr/blob/e61fc0a6c9e963e08df0f39a3bddb9a48bdacfe4/extendr-api/src/lib.rs#L815-L864

That's cool! You can make that an issue for a new PR or something. I'm not adding runtime tests in this PR.

**Comment by JosiahParry** [comment]
I'll adapt the rust tests into the extendrtests integration and push to this branch

File: extendr-api/src/wrapper/dataframe.rs [hunk 4]
```diff
@@ -7,21 +7,32 @@ pub trait IntoDataFrameRow<T> {
...
+
+impl<T> From<Dataframe<T>> for Robj {
+    fn from(value: Dataframe<T>) -> Self {
+        value.robj
+    }
+}
+
+impl<T> FromRobj<'_> for Dataframe<T> {
+    fn from_robj(robj: &Robj) -> std::result::Result<Self, &'static str> {
+        robj.try_into().map_err(|_| "expected a `data.frame`")
```
Comment (Ilia-Kosenkov) [inline extendr-api/src/wrapper/dataframe.rs:21]: Can you use `.or()` here since the error values is a (constant) static string?

Comment (CGMossa) [reply extendr-api/src/wrapper/dataframe.rs:21]: Sure! What's the difference?

File: extendr-api/src/wrapper/dataframe.rs [hunk 5]
```diff
@@ -4,6 +4,18 @@ pub trait IntoDataFrameRow<T> {
     fn into_dataframe(self) -> Result<Dataframe<T>>;
 }
 
+/// A representation of a typed data.frame
```
Comment (CGMossa) [inline extendr-api/src/wrapper/dataframe.rs:7]:
```suggestion
/// A representation of a typed `data.frame`
```

**Comment by JosiahParry** [comment]
Alright! I think this is in a place where it can be merged now once the next pass of CI is all green. 

Integration tests have been added at https://github.com/extendr/extendr/pull/723/files#diff-bf77ed62590e2a1349dae79b507ae6219c1c2d8e7e2f6c3b6dff0473c0133e49.


## Follow ups 

This PR has illuminated that `IntoDataFrameRow` is not implemented for `Robj` which means extendr impls cannot be in a `Dataframe<T>` . This is covered by https://github.com/extendr/extendr/issues/727

Additionally, `Dataframe<T>` can be used as a function argument parameter, however, it will not check `T` https://github.com/extendr/extendr/issues/729

**Review by JosiahParry — APPROVED** [review]

**Pull request merged**